### PR TITLE
CRIMRE-118 Use datastore for search filter

### PR DESCRIPTION
**Pull request by timpeat**
## Link to relevant ticket
[CRIMRE-118](https://dsdmoj.atlassian.net/browse/CRIMRE-118)
[CRIMRE-83](https://dsdmoj.atlassian.net/browse/CRIMRE-83)

## Notes for reviewer
Completes 83 as well as refactors previous search/filter tickets to use the datastore.

## How to manually test the feature

[CRIMRE-118]: https://dsdmoj.atlassian.net/browse/CRIMRE-118?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMRE-118]: https://dsdmoj.atlassian.net/browse/CRIMRE-118?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMRE-83]: https://dsdmoj.atlassian.net/browse/CRIMRE-83?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMRE-83]: https://dsdmoj.atlassian.net/browse/CRIMRE-83?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by timpeat — COMMENTED** [review]
Get next using datastore search to follow.

File: spec/system/get_next_application_spec.rb [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
     end
+  end
+
+  context 'when there is no next application' do
+    include_context 'when search results empty'
 
-    expect(page).to have_content('Your list')
-    expect(page).to have_content('There are no new applications to be reviewed')
+    it 'shows an error when there is no next application' do
+      3.times do
```
Comment (willmcb) [inline spec/system/get_next_application_spec.rb:19]: This is stubbed now with en empty response - so no need for 3.times

File: spec/system/search_applications/filter_by_caseworker_spec.rb [hunk 2]
```diff
@@ -1,95 +1,46 @@
...
 
-      select 'All assigned', from: 'filter-assigned-user-id-field'
-      click_button 'Search'
+      visit '/'
+      click_on 'Search'
     end
 
-    it 'shows all assigned applications' do
-      expect(page).to have_content('2 search result')
+    it 'can choose from "", "Unnassigned", "All assigned", and caseworkers' do
```
Comment (willmcb) [inline spec/system/search_applications/filter_by_caseworker_spec.rb:41]: Extra 'n' in Unassigned

**Review by willmcb — CHANGES_REQUESTED** [review]
👍 few small comments.

File: app/lib/types.rb [hunk 3]
```diff
@@ -51,6 +51,7 @@ module Types
   Uuid = String
   PhoneNumber = String
   Date = Date | JSON::Date
+  DateTime = DateTime | JSON::DateTime
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/lib/types.rb:54]: Requires PR namespace  Types in LaaCrimeSchemas. Will do separately.

File: app/models/application_search_filter.rb [hunk 4]
```diff
@@ -15,29 +15,47 @@ class ApplicationSearchFilter < ApplicationStruct
...
   end
 
   private
 
-  def users_with_assignments
-    User.where(
+  def assignees
+    return @assignees if @assignees
+
+    users = User.where(
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/models/application_search_filter.rb:38]: done

File: app/models/application_search_result.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+class ApplicationSearchResult < ApplicationStruct
+  attribute :applicant_name, Types::String
+  attribute :submitted_at, Types::DateTime
+  attribute :reference, Types::Integer
+  attribute :resource_id, Types::Uuid
+  attribute :status, Types::String
+
+  alias id resource_id
+
+  def days_passed
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/models/application_search_result.rb:10]: Ticket added to backlog.

**Review by timpeat — COMMENTED** [review]
@arthurashman @willmcb Ready for another review.

**Review by willmcb — APPROVED** [review]
👍

**Pull request merged**